In [1]:
import pandas as pd
import numpy as np

# Assessing
Use the space below to explore `all_alpha_08.csv` and `all_alpha_18.csv` to answer the quiz questions below.
- number of samples in each dataset
- number of columns in each dataset
- duplicate rows in each dataset
- datatypes of columns
- features with missing values
- number of non-null unique values for features in each dataset
- what those unique values are and counts for each

In [2]:
fuel_08_df = pd.read_csv('all_alpha_08.csv')
fuel_18_df = pd.read_csv('all_alpha_18.csv')

In [3]:
print(fuel_08_df.shape)
print(fuel_18_df.shape)

(2404, 18)
(1611, 18)


There are 18 colummns and 1611 samples in the 2018 dataset


In [4]:
print(fuel_08_df.duplicated().sum())
print(fuel_18_df.duplicated().sum())

25
0


In [5]:
fuel_08_df.isnull().sum()

Model                     0
Displ                     0
Cyl                     199
Trans                   199
Drive                    93
Fuel                      0
Sales Area                0
Stnd                      0
Underhood ID              0
Veh Class                 0
Air Pollution Score       0
FE Calc Appr            199
City MPG                199
Hwy MPG                 199
Cmb MPG                 199
Unadj Cmb MPG           199
Greenhouse Gas Score    199
SmartWay                  0
dtype: int64

In [6]:
fuel_18_df.isnull().sum()

Model                   0
Displ                   2
Cyl                     2
Trans                   0
Drive                   0
Fuel                    0
Cert Region             0
Stnd                    0
Stnd Description        0
Underhood ID            0
Veh Class               0
Air Pollution Score     0
City MPG                0
Hwy MPG                 0
Cmb MPG                 0
Greenhouse Gas Score    0
SmartWay                0
Comb CO2                0
dtype: int64

In [7]:
fuel_08_df.dtypes

Model                    object
Displ                   float64
Cyl                      object
Trans                    object
Drive                    object
Fuel                     object
Sales Area               object
Stnd                     object
Underhood ID             object
Veh Class                object
Air Pollution Score      object
FE Calc Appr             object
City MPG                 object
Hwy MPG                  object
Cmb MPG                  object
Unadj Cmb MPG           float64
Greenhouse Gas Score     object
SmartWay                 object
dtype: object

In [8]:
fuel_18_df.dtypes

Model                    object
Displ                   float64
Cyl                     float64
Trans                    object
Drive                    object
Fuel                     object
Cert Region              object
Stnd                     object
Stnd Description         object
Underhood ID             object
Veh Class                object
Air Pollution Score       int64
City MPG                 object
Hwy MPG                  object
Cmb MPG                  object
Greenhouse Gas Score      int64
SmartWay                 object
Comb CO2                 object
dtype: object

In [9]:
fuel_08_df.nunique()

Model                   436
Displ                    47
Cyl                       8
Trans                    14
Drive                     2
Fuel                      5
Sales Area                3
Stnd                     12
Underhood ID            343
Veh Class                 9
Air Pollution Score      13
FE Calc Appr              2
City MPG                 39
Hwy MPG                  43
Cmb MPG                  38
Unadj Cmb MPG           721
Greenhouse Gas Score     20
SmartWay                  2
dtype: int64

In [10]:
fuel_18_df.nunique()

Model                   367
Displ                    36
Cyl                       7
Trans                    26
Drive                     2
Fuel                      5
Cert Region               2
Stnd                     19
Stnd Description         19
Underhood ID            230
Veh Class                 9
Air Pollution Score       6
City MPG                 58
Hwy MPG                  62
Cmb MPG                  57
Greenhouse Gas Score     10
SmartWay                  3
Comb CO2                299
dtype: int64

How do the Cyl columns in the 2008 and 2018 datasets differ?
- The differ in Data types, Format and Number of unique values

In [11]:
fuel_08_df.Fuel.unique()

array(['Gasoline', 'ethanol/gas', 'ethanol', 'diesel', 'CNG'],
      dtype=object)

In [12]:
fuel_18_df.Fuel.unique()

array(['Gasoline', 'Gasoline/Electricity', 'Diesel', 'Ethanol/Gas',
       'Electricity'], dtype=object)

# Cleaning Column labels


In [13]:
fuel_08_df.head(1)

Model  Displ      Cyl    Trans Drive      Fuel Sales Area Stnd  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline         CA   U2   

   Underhood ID Veh Class Air Pollution Score FE Calc Appr City MPG Hwy MPG  \
0  8HNXT03.7PKR       SUV                   7          Drv       15      20   

  Cmb MPG  Unadj Cmb MPG Greenhouse Gas Score SmartWay  
0      17        22.0527                    4       no

In [14]:
fuel_18_df.head(1)

Model  Displ  Cyl       Trans Drive      Fuel Cert Region    Stnd  \
0  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline          FA  T3B125   

         Stnd Description  Underhood ID  Veh Class  Air Pollution Score  \
0  Federal Tier 3 Bin 125  JHNXT03.5GV3  small SUV                    3   

  City MPG Hwy MPG Cmb MPG  Greenhouse Gas Score SmartWay Comb CO2  
0       20      28      23                     5       No      386

### Drop Extraneous Columns

In [15]:
# drop columns from 2008 dataset
fuel_08_df.drop(['Stnd', 'Underhood ID', 'FE Calc Appr', 'Unadj Cmb MPG'], axis=1, inplace=True)

# confirm changes
fuel_08_df.head(1)

Model  Displ      Cyl    Trans Drive      Fuel Sales Area Veh Class  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline         CA       SUV   

  Air Pollution Score City MPG Hwy MPG Cmb MPG Greenhouse Gas Score SmartWay  
0                   7       15      20      17                    4       no

In [16]:
# drop columns from 2008 dataset
fuel_18_df.drop(['Stnd', 'Stnd Description', 'Underhood ID', 'Comb CO2'], axis=1, inplace=True)

# confirm changes
fuel_18_df.head(1)

Model  Displ  Cyl       Trans Drive      Fuel Cert Region  Veh Class  \
0  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline          FA  small SUV   

   Air Pollution Score City MPG Hwy MPG Cmb MPG  Greenhouse Gas Score SmartWay  
0                    3       20      28      23                     5       No

### Rename Columns

In [17]:
# rename Sales Area to Cert Region in 2008 dataset
fuel_08_df.rename(columns={'Sales Area':'Cert Region'}, inplace=True)

# confirm changes
fuel_08_df.head(1)

Model  Displ      Cyl    Trans Drive      Fuel Cert Region Veh Class  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline          CA       SUV   

  Air Pollution Score City MPG Hwy MPG Cmb MPG Greenhouse Gas Score SmartWay  
0                   7       15      20      17                    4       no

In [18]:
# replace spaces with underscores and lowercase labels for 2008 dataset
fuel_08_df.rename(columns=lambda x: x.strip().lower().replace(" ", "_"), inplace=True) # Strip method removes spaces(by default) at the beginning and end

# confirm changes
fuel_08_df.head(1)

model  displ      cyl    trans drive      fuel cert_region veh_class  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline          CA       SUV   

  air_pollution_score city_mpg hwy_mpg cmb_mpg greenhouse_gas_score smartway  
0                   7       15      20      17                    4       no

In [19]:
# replace spaces with underscores and lowercase labels for 2008 dataset
fuel_18_df.rename(columns=lambda x: x.strip().lower().replace(" ", "_"), inplace=True)

# confirm changes
fuel_18_df.head(1)

model  displ  cyl       trans drive      fuel cert_region  veh_class  \
0  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline          FA  small SUV   

   air_pollution_score city_mpg hwy_mpg cmb_mpg  greenhouse_gas_score smartway  
0                    3       20      28      23                     5       No

In [20]:
# confirm column labels for 2008 and 2018 datasets are identical
fuel_08_df.columns == fuel_18_df.columns

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True])

In [21]:
# make sure they're all identical like this
(fuel_08_df.columns == fuel_18_df.columns).all()

True

# Filter, Drop Nulls, Dedupe
1. Filter
For consistency, only compare cars certified by California standards. Filter both datasets using query to select only rows where cert_region is CA. Then, drop the cert_region columns, since it will no longer provide any useful information (we'll know every value is 'CA').

2. Drop Nulls
Drop any rows in both datasets that contain missing values.

3. Dedupe
Drop any duplicate rows in both datasets.

In [22]:
fuel_08_df.shape

(2404, 14)

In [23]:
fuel_18_df.shape

(1611, 14)

## Filter by Certification Region

In [24]:
df_08 = fuel_08_df.query('cert_region == "CA"')
df_18 = fuel_18_df.query('cert_region == "CA"')

In [25]:
# confirm only certification region is California
df_08['cert_region'].unique()

array(['CA'], dtype=object)

In [26]:
# confirm only certification region is California
df_18['cert_region'].unique()

array(['CA'], dtype=object)

In [27]:
# drop certification region columns form both datasets
df_08.drop('cert_region', axis=1, inplace=True)
df_18.drop('cert_region', axis=1, inplace=True)

C:\Users\hp\AppData\Local\Temp\ipykernel_13520\3207701445.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_08.drop('cert_region', axis=1, inplace=True)
C:\Users\hp\AppData\Local\Temp\ipykernel_13520\3207701445.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_18.drop('cert_region', axis=1, inplace=True)


In [28]:
df_08.shape

(1084, 13)

In [29]:
df_08.shape

(1084, 13)

## Drop Rows with Missing Values

In [30]:
df_08.isnull().sum()

model                    0
displ                    0
cyl                     75
trans                   75
drive                   37
fuel                     0
veh_class                0
air_pollution_score      0
city_mpg                75
hwy_mpg                 75
cmb_mpg                 75
greenhouse_gas_score    75
smartway                 0
dtype: int64

In [31]:
df_18.isnull().sum()

model                   0
displ                   1
cyl                     1
trans                   0
drive                   0
fuel                    0
veh_class               0
air_pollution_score     0
city_mpg                0
hwy_mpg                 0
cmb_mpg                 0
greenhouse_gas_score    0
smartway                0
dtype: int64

In [32]:
# drop rows with any null values in both datasets
df_08.dropna(inplace=True)
df_18.dropna(inplace=True)

C:\Users\hp\AppData\Local\Temp\ipykernel_13520\728496535.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_08.dropna(inplace=True)
C:\Users\hp\AppData\Local\Temp\ipykernel_13520\728496535.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_18.dropna(inplace=True)


## Dedupe Data

In [33]:
# print number of duplicates in 2008 and 2018 datasets
print(df_08.duplicated().sum())
print(df_18.duplicated().sum())

23
3


In [34]:
# drop duplicates in both datasets
df_08.drop_duplicates(inplace=True)
df_18.drop_duplicates(inplace=True)

C:\Users\hp\AppData\Local\Temp\ipykernel_13520\1305320245.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_08.drop_duplicates(inplace=True)
C:\Users\hp\AppData\Local\Temp\ipykernel_13520\1305320245.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_18.drop_duplicates(inplace=True)


In [35]:
# print number of duplicates again to confirm dedupe - should both be 0
print(df_08.duplicated().sum())
print(df_18.duplicated().sum())

0
0


## Fixing cyl Data Type

In [36]:
df_08['cyl'].value_counts()

(6 cyl)     409
(4 cyl)     283
(8 cyl)     199
(5 cyl)      48
(12 cyl)     30
(10 cyl)     14
(2 cyl)       2
(16 cyl)      1
Name: cyl, dtype: int64

Read [this](https://stackoverflow.com/questions/35376387/extract-int-from-string-in-pandas) to help you extract ints from strings in Pandas for the next step.

In [37]:
# Extract int from strings in the 2008 cyl column
df_08['cyl'] = df_08['cyl'].str.extract('(\d+)').astype(int)

C:\Users\hp\AppData\Local\Temp\ipykernel_13520\2003834998.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_08['cyl'] = df_08['cyl'].str.extract('(\d+)').astype(int)


In [38]:
# Check value counts for 2008 cyl column again to confirm the change
df_08['cyl'].value_counts()

6     409
4     283
8     199
5      48
12     30
10     14
2       2
16      1
Name: cyl, dtype: int64

In [39]:
# convert 2018 cyl column to int
df_18['cyl'] = df_18['cyl'].astype(int)

C:\Users\hp\AppData\Local\Temp\ipykernel_13520\1840797855.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_18['cyl'] = df_18['cyl'].astype(int)


# Fixing `air_pollution_score` Data Type
- 2008: convert string to float
- 2018: convert int to float


In [40]:
# try using Pandas to_numeric or astype function to convert the
# 2008 air_pollution_score column to float -- this won't work
df_08['air_pollution_score'] = df_08['air_pollution_score'].astype(float)

ValueError: could not convert string to float: '6/4'

## Figuring out the issue
Looks like this isn't going to be as simple as converting the datatype. According to the error above, the air pollution score value in one of the rows is "6/4" - let's check it out.

In [ ]:
df_08[df_08['air_pollution_score'] == '6/4']

# It's not just the air pollution score!
The mpg columns and greenhouse gas scores also seem to have the same problem - maybe that's why these were all saved as strings! According to [this link](http://www.fueleconomy.gov/feg/findacarhelp.shtml#airPollutionScore), which I found from the PDF documentation:

    "If a vehicle can operate on more than one type of fuel, an estimate is provided for each fuel type."
    
Ohh.. so all vehicles with more than one fuel type, or hybrids, like the one above (it uses ethanol AND gas) will have a string that holds two values - one for each. This is a little tricky

In [41]:
# First, let's get all the hybrids in 2008
hb_08 = df_08[df_08['fuel'].str.contains('/')]
hb_08

model  displ  cyl    trans drive         fuel  veh_class  \
1550  MERCEDES-BENZ C300    3.0    6  Auto-L7   2WD  ethanol/gas  small car   

     air_pollution_score city_mpg hwy_mpg cmb_mpg greenhouse_gas_score  \
1550                 6/4    13/18   19/25   15/21                  7/6   

     smartway  
1550       no

Looks like this dataset only has one! Now for 2018

In [54]:
# hybrids in 2018
hb_18 = df_18[df_18['fuel'].str.contains('/')]
hb_18.head()

model  displ  cyl       trans drive                  fuel  \
108          BMW 330e    2.0    4  SemiAuto-8   2WD  Gasoline/Electricity   
160          BMW 530e    2.0    4  SemiAuto-8   2WD  Gasoline/Electricity   
162          BMW 530e    2.0    4  SemiAuto-8   4WD  Gasoline/Electricity   
188          BMW 740e    2.0    4  SemiAuto-8   4WD  Gasoline/Electricity   
382  CHEVROLET Impala    3.6    6  SemiAuto-6   2WD           Ethanol/Gas   

     veh_class  air_pollution_score city_mpg hwy_mpg cmb_mpg  \
108  small car                    3    28/66   34/78   30/71   
160  small car                    7    27/70   31/75   29/72   
162  small car                    7    27/66   31/68   28/67   
188  large car                    3    25/62   29/68   27/64   
382  large car                    5    14/18   20/28   16/22   

     greenhouse_gas_score smartway  
108                    10      Yes  
160                    10    Elite  
162                    10    Elite  
188                     9      Yes  
382                     4       No

We're going to take each hybrid row and split them into two new rows - one with values for the first fuel type (values before the "/"), and the other with values for the second fuel type (values after the "/"). Let's separate them with two dataframes!

In [43]:
# create two copies of the 2008 hybrids dataframe
df1 = hb_08.copy()  # data on first fuel type of each hybrid vehicle
df2 = hb_08.copy()  # data on second fuel type of each hybrid vehicle

# Each one should look like this
df1    

model  displ  cyl    trans drive         fuel  veh_class  \
1550  MERCEDES-BENZ C300    3.0    6  Auto-L7   2WD  ethanol/gas  small car   

     air_pollution_score city_mpg hwy_mpg cmb_mpg greenhouse_gas_score  \
1550                 6/4    13/18   19/25   15/21                  7/6   

     smartway  
1550       no

For this next part, I'm going to use Pandas' apply function. See the docs [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.apply.html).

In [44]:
# columns to split by "/"
split_columns = ['fuel', 'air_pollution_score', 'city_mpg', 'hwy_mpg', 'cmb_mpg', 'greenhouse_gas_score']

# apply split function to each column of each dataframe copy
for c in split_columns:
    df1[c] = df1[c].apply(lambda x: x.split("/")[0])
    df2[c] = df2[c].apply(lambda x: x.split("/")[1])

In [45]:
# this dataframe holds info for the FIRST fuel type of the hybrid
# aka the values before the "/"s
df1

model  displ  cyl    trans drive     fuel  veh_class  \
1550  MERCEDES-BENZ C300    3.0    6  Auto-L7   2WD  ethanol  small car   

     air_pollution_score city_mpg hwy_mpg cmb_mpg greenhouse_gas_score  \
1550                   6       13      19      15                    7   

     smartway  
1550       no

In [46]:
# this dataframe holds info for the SECOND fuel type of the hybrid
# aka the values before the "/"s
df2

model  displ  cyl    trans drive fuel  veh_class  \
1550  MERCEDES-BENZ C300    3.0    6  Auto-L7   2WD  gas  small car   

     air_pollution_score city_mpg hwy_mpg cmb_mpg greenhouse_gas_score  \
1550                   4       18      25      21                    6   

     smartway  
1550       no

In [50]:
# combine dataframes to add to the original dataframe
new_rows = pd.concat([df1,df2])

# now we have separate rows for each fuel type of each vehicle!
new_rows

model  displ  cyl    trans drive     fuel  veh_class  \
1550  MERCEDES-BENZ C300    3.0    6  Auto-L7   2WD  ethanol  small car   
1550  MERCEDES-BENZ C300    3.0    6  Auto-L7   2WD      gas  small car   

     air_pollution_score city_mpg hwy_mpg cmb_mpg greenhouse_gas_score  \
1550                   6       13      19      15                    7   
1550                   4       18      25      21                    6   

     smartway  
1550       no  
1550       no

In [ ]:
# drop the original hybrid rows
df_08.drop(hb_08.index, inplace=True)

# add in our newly separated rows
df_08 = df_08.append(new_rows, ignore_index=True)

In [52]:
# check that all the original hybrid rows with "/"s are gone
df_08[df_08['fuel'].str.contains('/')]

Empty DataFrame
Columns: [model, displ, cyl, trans, drive, fuel, veh_class, air_pollution_score, city_mpg, hwy_mpg, cmb_mpg, greenhouse_gas_score, smartway]
Index: []

In [53]:
df_08.shape

(987, 13)

## Repeat this process for the 2018 dataset

In [57]:
# create two copies of the 2018 hybrids dataframe, hb_18
df1 = hb_18.copy()
df2 = hb_18.copy()

### Split values for `fuel`, `city_mpg`, `hwy_mpg`, `cmb_mpg`
You don't need to split for `air_pollution_score` or `greenhouse_gas_score` here because these columns are already ints in the 2018 dataset.

In [58]:
# list of columns to split
split_columns = ['fuel', 'city_mpg', 'hwy_mpg', 'cmb_mpg']

# apply split function to each column of each dataframe copy
for c in split_columns:
    df1[c] = df1[c].apply(lambda x: x.split("/")[0])
    df2[c] = df2[c].apply(lambda x: x.split("/")[1])

In [60]:
# append the two dataframes
new_rows = pd.concat([df1,df2])

# drop each hybrid row from the original 2018 dataframe
# do this by using Pandas drop function with hb_18's index
df_18.drop(hb_18.index, inplace=True)

# append new_rows to df_18

df_18 = pd.concat([df_18,new_rows],ignore_index=True)


C:\Users\hp\AppData\Local\Temp\ipykernel_13520\2013587141.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_18.drop(hb_18.index, inplace=True)


In [61]:
# check that they're gone
df_18[df_18['fuel'].str.contains('/')]

Empty DataFrame
Columns: [model, displ, cyl, trans, drive, fuel, veh_class, air_pollution_score, city_mpg, hwy_mpg, cmb_mpg, greenhouse_gas_score, smartway]
Index: []

In [62]:
df_18.shape

(832, 13)

### Now we can comfortably continue the changes needed for `air_pollution_score`! Here they are again:
- 2008: convert string to float
- 2018: convert int to float

In [63]:
# convert string to float for 2008 air pollution column
df_08.air_pollution_score = df_08.air_pollution_score.astype(float)

In [64]:
# convert int to float for 2018 air pollution column
df_18.air_pollution_score = df_18.air_pollution_score.astype(float)

## Fix `city_mpg`, `hwy_mpg`, `cmb_mpg` datatypes
    2008 and 2018: convert string to float


In [65]:
# convert mpg columns to floats
mpg_columns = ['city_mpg', 'hwy_mpg', 'cmb_mpg']
for c in mpg_columns:
    df_18[c] = df_18[c].astype(float)
    df_08[c] = df_08[c].astype(float)

## Fix `greenhouse_gas_score` datatype
    2008: convert from float to int

In [66]:
# convert from float to int
df_08['greenhouse_gas_score'] = df_08['greenhouse_gas_score'].astype(int)

## All the dataypes are now fixed! Lets take one last check to confirm all the changes.

In [67]:
df_08.dtypes

model                    object
displ                   float64
cyl                       int32
trans                    object
drive                    object
fuel                     object
veh_class                object
air_pollution_score     float64
city_mpg                float64
hwy_mpg                 float64
cmb_mpg                 float64
greenhouse_gas_score      int32
smartway                 object
dtype: object

In [68]:
df_18.dtypes

model                    object
displ                   float64
cyl                       int32
trans                    object
drive                    object
fuel                     object
veh_class                object
air_pollution_score     float64
city_mpg                float64
hwy_mpg                 float64
cmb_mpg                 float64
greenhouse_gas_score      int64
smartway                 object
dtype: object

In [70]:
df_08.dtypes == df_18.dtypes

model                    True
displ                    True
cyl                      True
trans                    True
drive                    True
fuel                     True
veh_class                True
air_pollution_score      True
city_mpg                 True
hwy_mpg                  True
cmb_mpg                  True
greenhouse_gas_score    False
smartway                 True
dtype: bool